File: electron/toolCallUtils.js
/**
 * 规范化 tool calls（过滤无效项，统一字段形状）。
 * 可复用于流式合并后的调用列表与非流式 message.tool_calls。
 */
function normalizeToolCallsList(list) {
  const arr = Array.isArray(list) ? list : []
  return arr
    .filter((tc) => tc && (tc.id || '').trim() && (tc.function?.name || '').trim())
    .map((tc) => ({
      id: String(tc.id || '').trim(),
      type: tc.type || 'function',
      function: {
        name: (tc.function?.name || '').trim(),
        arguments: typeof tc.function?.arguments === 'string' ? tc.function.arguments : '',
      },
    }))
}

module.exports = {
  normalizeToolCallsList,
}

File: backend/utils/tool_call_utils.py
"""
Tool call normalisation — port of electron/toolCallUtils.js.
"""

from __future__ import annotations

from typing import Any


def normalize_tool_calls_list(calls: list[dict[str, Any]] | None) -> list[dict]:
    """Filter invalid entries and unify field shapes."""
    if not calls:
        return []
    result: list[dict] = []
    for tc in calls:
        if not tc:
            continue
        tc_id = (tc.get("id") or "").strip()
        fn = tc.get("function") or {}
        name = (fn.get("name") or "").strip()
        if not tc_id or not name:
            continue
        result.append({
            "id": tc_id,
            "type": tc.get("type", "function"),
            "function": {
                "name": name,
                "arguments": fn.get("arguments", "") if isinstance(fn.get("arguments"), str) else "",
            },
        })
    return result
